sklearn/cluster/_hdbscan/hdbscan.py

```diff
@@ -0,0 +1,751 @@
+                X = X[finite_index]
+        elif issparse(X):
+            # Handle sparse precomputed distance matrices separately
+            X = self._validate_data(
+                X,
+                accept_sparse=["csr", "lil"],
+                dtype=np.float64,
+            )
+        else:
+            # Only non-sparse, precomputed distance matrices are handled here
+            # and thereby allowed to contain numpy.inf for missing distances
+
+            # Perform data validation after removing infinite values (numpy.inf)
+            # from the given distance matrix.
+            X = self._validate_data(X, force_all_finite=False, dtype=np.float64)
+            if np.isnan(X).any():
+                # TODO: Support np.nan in Cython implementation for sparse
+                # HDBSCAN
+                raise ValueError("np.nan values found in precomputed-sparse")
+        if X.shape[0] == 1:
+            raise ValueError("n_samples=1 while HDBSCAN requires more than one sample")
+        self.n_features_in_ = X.shape[1]
+        self._min_samples = (
+            self.min_cluster_size if self.min_samples is None else self.min_samples
+        )
+
+        if self._min_samples > X.shape[0]:
+            raise ValueError(
+                f"min_samples ({self._min_samples}) must be at most the number of"
+                f" samples in X ({X.shape[0]})"
+            )
+        mst_func = None
+        kwargs = dict(
+            X=X,
+            algo="kd_tree",
+            min_samples=self._min_samples,
+            alpha=self.alpha,
+            metric=self.metric,
+            leaf_size=self.leaf_size,
+            n_jobs=self.n_jobs,
+            **self._metric_params,
+        )
+        if self.algorithm == "kdtree" and self.metric not in KDTree.valid_metrics:
+            raise ValueError(
+                f"{self.metric} is not a valid metric for a KDTree-based algorithm."
+                " Please select a different metric."
+            )
+        elif self.algorithm == "balltree" and self.metric not in BallTree.valid_metrics:
+            raise ValueError(
+                f"{self.metric} is not a valid metric for a BallTree-based algorithm."
+                " Please select a different metric."
+            )
+
+        if self.algorithm != "auto":
+            if (
+                self.metric != "precomputed"
+                and issparse(X)
+                and self.algorithm != "brute"
+            ):
+                raise ValueError("Sparse data matrices only support algorithm `brute`.")
+
+            if self.algorithm == "brute":
+                mst_func = _hdbscan_brute
+                kwargs["copy"] = self.copy
+                for key in ("algo", "leaf_size"):
+                    kwargs.pop(key, None)
+            elif self.algorithm == "kdtree":
+                mst_func = _hdbscan_prims
+            elif self.algorithm == "balltree":
```

Comment:
I think overriding or deleting items in `kwargs` here is a bit confusing. Can we build `kwargs` instead? i.e:

```python
# do not include "algo" or "leaf_size" when defining `kwargs`
kwargs = {
    X=X,
    min_samples=self._min_samples
}

if self.algorithm != "auto":
    if self.algorithm == "brute":
        ...
    elif self.algorithm == "kdtree":
        kwargs["algo"] = "kd_tree"
        kwargs["leaf_size"] = self.leaf_size
    elif self.algorithm == "balltree":
        kwargs["algo"] = "ball_tree"
        kwargs["leaf_size"] = self.leaf_size
```

This way "algo" and "leaf_size" is set when it is needed.